core/drivers/crypto/caam/cipher/caam_cipher.c

```diff
@@ -669,157 +568,207 @@ static TEE_Result do_update_streaming(struct drvcrypt_cipher_update *dupdate)
-			srcbuf.data = dupdate->src.data;
-			srcbuf.length = size_todo;
-			srcbuf.paddr = psrc;
-
-			dstbuf.data = dst_align.data;
-			dstbuf.length = size_todo;
-			dstbuf.paddr = dst_align.paddr;
-			dstbuf.nocache = dst_align.nocache;
-
-			retstatus =
-				caam_cipher_block(ctx, true, NEED_KEY1,
-						  ctx->encrypt, &srcbuf,
-						  &dstbuf, CIPHER_BLOCK_NONE);
-		}
+	size_done = size_todo;
+	dupdate->dst.length = 0;
+	for (offset = 0; size_todo;
+	     offset += size_done, size_todo -= size_done) {
+		CIPHER_TRACE("Do input %zu bytes (%zu), offset %zu", size_done,
+			     size_todo, offset);
+
+		size_done = size_todo;
+		ret = caam_dmaobj_sgtbuf_inout_build(&src, &dst, &size_done,
+						     offset,
+						     ctx->alg->size_block);
+		if (ret)
+			goto end_streaming;
+
+		retstatus = caam_cipher_block(ctx, true, NEED_KEY1,
+					      ctx->encrypt, &src, &dst);
 
 		if (retstatus != CAAM_NO_ERROR) {
-			ret = TEE_ERROR_GENERIC;
-			goto out;
+			ret = caam_status_to_tee_result(retstatus);
+			goto end_streaming;
 		}
 
-		CIPHER_DUMPBUF("Source", dupdate->src.data,
-			       dupdate->src.length - size_topost);
-		CIPHER_DUMPBUF("Result", dst_align.data,
-			       dupdate->dst.length - size_topost);
+		dupdate->dst.length += caam_dmaobj_copy_to_orig(&dst);
 	}
 
+	CIPHER_DUMPBUF("Source", dupdate->src.data, dupdate->src.length);
+	CIPHER_DUMPBUF("Result", dupdate->dst.data, dupdate->dst.length);
+
+end_streaming_post:
 	if (size_topost) {
+		caam_dmaobj_free(&src);
+		caam_dmaobj_free(&dst);
+		CIPHER_TRACE("Save input data %zu bytes (done %zu) - off %zu",
+			     size_topost, size_inmade, offset);
+
+		size_todo = size_topost + ctx->blockbuf.filled;
+
+		/*
+		 * Prepare the destination DMA Object:
+		 *  - Use given destination parameter bytes to return
+		 *  - If the previous operation saved data, use a trash
+		 *    buffer to do the operation but not use unneeded data.
+		 */
+		ret = caam_dmaobj_init_output(&dst,
+					      dupdate->dst.data + size_inmade,
+					      size_topost, size_topost);
+		if (ret)
+			goto end_streaming;
+
+		ret = caam_dmaobj_prepare(NULL, &dst, ctx->alg->size_block);
+		if (ret)
+			goto end_streaming;
+
+		if (ctx->blockbuf.filled) {
+			/*
+			 * Because there are some bytes to trash, use
+			 * a block buffer that will be added to the
+			 * destination SGT/Buffer structure to do the
+			 * cipher operation.
+			 */
+			ret = caam_alloc_align_buf(&trash_bck.buf,
+						   ctx->blockbuf.filled);
+			if (ret != CAAM_NO_ERROR) {
+				CIPHER_TRACE("Allocation Trash Block error");
+				goto end_streaming;
+			}
+			trash_bck.filled = ctx->blockbuf.filled;
+
+			ret = caam_dmaobj_add_first_block(&dst, &trash_bck);
+			if (ret)
+				goto end_streaming;
+		}
+
+		/*
+		 * Save the input data in the block buffer for next operation
+		 * and prepare the source DMA Object with the overall saved
+		 * data to generate destination bytes.
+		 */
 		struct caambuf cpysrc = {
```

Comment:
Please keep declarations of variables at the start of a block.

Reply:
Ok